contracts/dao/neutron-flashloans/src/contract.rs

```diff
@@ -0,0 +1,403 @@
+fn get_pre_loan_balances(
+    deps: &DepsMut,
+    source: Addr,
+    requested_amount: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
+        address: source.to_string(),
+    };
+    // Get the response (all balances)
+    let all_balances_response: AllBalanceResponse =
+        deps.querier.query(&all_balances_query.into())?;
+
+    // Filter all balances leaving only the balances of the requested coins
+    let mut pre_loan_balances: Vec<Coin> = vec![];
+    for requested_coin in requested_amount {
+        // Look for the requested coin in the source balances, AND check that the source
+        // has enough of the requested coin.
+        let maybe_source_coin = all_balances_response
+            .amount
+            .iter()
+            .find(|x| x.denom == requested_coin.denom && requested_coin.amount.le(&x.amount));
+
+        // If the source doesn't have (enough of) the requested coin, return an error
+        if maybe_source_coin.is_none() {
+            return Err(ContractError::InsufficientFunds {
+                denom: requested_coin.denom,
+            });
+        }
+
+        let source_coin = maybe_source_coin.unwrap();
+        pre_loan_balances.push(source_coin.clone())
+    }
+
+    Ok(pre_loan_balances)
+}
+
+/// Calculates the fee by multiplying each of the requested assets by fee_rate.
+fn calculate_fee(
+    requested_amount: Vec<Coin>,
+    fee_rate: Decimal,
+) -> Result<Vec<Coin>, ContractError> {
+    let mut fee: Vec<Coin> = Vec::with_capacity(requested_amount.len());
+    for coin in requested_amount {
+        let coin_fee = Coin::new((fee_rate * coin.amount).u128(), coin.denom);
+        fee.push(coin_fee)
+    }
+
+    Ok(fee)
+}
+
+// Sums the pre_loan_balances with the fee.
+// WARNING: this function assumes that the input vectors are of the same length,
+// and that the order of the denoms is the same.
+fn calculate_expected_balances(
+    pre_loan_balances: Vec<Coin>,
+    fee: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    let mut expected_balances: Vec<Coin> = Vec::with_capacity(pre_loan_balances.len());
+    for (index, coin) in pre_loan_balances.iter().enumerate() {
+        expected_balances.push(Coin::new(
+            coin.amount.checked_add(fee[index].amount)?.u128(),
+            coin.denom.clone(),
+        ))
+    }
+
+    Ok(expected_balances)
+}
+
+/// A simple function to build the (stargate -> authz -> bank) /cosmos.bank.v1beta1.MsgSend message.
+fn get_stargate_authz_bank_send_msg(
+    env: Env,
+    config: Config,
+    borrower: Addr,
+    amount: Vec<Coin>,
+) -> CosmosMsg<NeutronMsg> {
+    // First we create the bank MsgSend
+    let bank_send_msg = MsgSend {
+        from_address: config.source.to_string(),
+        to_address: borrower.to_string(),
+        amount: amount
+            .iter()
+            .map(|x| ProtoCoin {
+                denom: x.clone().denom,
```

Comment:
why clone the whole coin instance?
```suggestion
                denom: x.denom.clone(),
```